### ISSUE TITLE
[FEA] Enable dynamic "has_nulls" for row operators 

### ISSUE THREAD

**Issue by jrhemstad** [rapidsai/cudf#9580]
**Is your feature request related to a problem? Please describe.**

The [`row_lexicographic_comparator`](https://github.com/rapidsai/cudf/blob/8a5adad1cbb6851b2a7b658ac4218e749210803c/cpp/include/cudf/table/row_operators.cuh#L333-L334) and [`row_equality_comparator`](https://github.com/rapidsai/cudf/blob/8a5adad1cbb6851b2a7b658ac4218e749210803c/cpp/include/cudf/table/row_operators.cuh#L225-L226) have a `has_nulls` non-type template parameter to allow optimizing away the more expensive null handling code.

While this is nice in cases where the null logic actually impacts performance, there are times where it doesn't. In the cases where it doesn't matter for performance, it needlessly increases compile time by having two versions of a kernel (`has_nulls==true` and `has_nulls==false`). 

We should make it so this information can be dynamic in addition to static.

**Describe the solution you'd like**

Similar to `optional_iterator` change the `has_nulls` to have three options:
```
row_equality_comparator<HAS_NULLS::YES>{...}
row_equality_comparator<HAS_NULLS::NO>{...}
row_equality_comparator<HAS_NULLS::DYNAMIC>{..., has_nulls}; // runtime value for has_nulls
```

The implementation should be specialized to ignore the `has_nulls` dynamic value for the two static cases.

**Comment by davidwendt** [comment]
I had already started looking at an approach for this awhile back. Summary of my findings are in https://github.com/rapidsai/cudf/issues/6952

**Comment by davidwendt** [comment]
Closed by #9623
